### ISSUE TITLE
Cached results can mask integration tests failures

### ISSUE THREAD

**Issue by mr-salty** [googleapis/google-cloud-cpp#5460]
This issue was actually caused by #5445 but as @coryan noted in #5456 the CI tests continued to use the cached results.

I will disable this test temporarily until we can fix it.

build failure:

```
[ RUN      ] ObjectResumableWriteIntegrationTest.WriteResumeFinalizedUpload
google/cloud/storage/tests/object_resumable_write_integration_test.cc:221: Failure
Value of: os.IsOpen()
  Actual: true
Expected: false
google/cloud/storage/tests/object_resumable_write_integration_test.cc:223: Failure
Value of: os.metadata()
Expected: is OK
Actual: default [UNKNOWN]
[  FAILED  ] ObjectResumableWriteIntegrationTest.WriteResumeFinalizedUpload (908 ms)
```

**Comment by coryan** [comment]
After looking at it some more, it is worse and better: it is a problem with cached results, but the build in question uses **production** (no the emulator) for the storage tests.  The successful runs are indeed cached, but I now suspects the service changed behavior.

**Comment by vnvo2409** [comment]
> but the build in question uses production (no the emulator) for the storage tests

So I suppose there is nothing to do with the new emulator.

> The successful runs are indeed cached

Bazel has an option [`--cache_test_results`](https://docs.bazel.build/versions/master/user-manual.html#flag--cache_test_results). I think we could force `--cache_test_results=no` in an integration test and a production test ( Maybe occasionally only, like 1 specific day in every month, with a little `bash` trick )

> build failure:
```
[ RUN      ] ObjectResumableWriteIntegrationTest.WriteResumeFinalizedUpload
google/cloud/storage/tests/object_resumable_write_integration_test.cc:221: Failure
Value of: os.IsOpen()
  Actual: true
Expected: false
google/cloud/storage/tests/object_resumable_write_integration_test.cc:223: Failure
Value of: os.metadata()
Expected: is OK
Actual: default [UNKNOWN]
[  FAILED  ] ObjectResumableWriteIntegrationTest.WriteResumeFinalizedUpload (908 ms)
```
I've already implemented a feature for dumping the `testbench.log` if there is any failed test. So please add the link to the `build.log` so we could have a better insight about what happend inside the emulator ( I think we do not need it for this issue though).

**Comment by mr-salty** [comment]
The caching saves a lot of resources in general; I don't know if there is a practical way to invalidate the cache if the backend changes, though (and that is likely a fairly rare problem).  #5462 proposes a solution for emulator changes (which wouldn't have helped here).

But, the idea of periodically invalidating the cache is good, maybe we can do an overnight build for that purpose.

**Comment by coryan** [comment]
> But, the idea of periodically invalidating the cache is good, maybe we can do an overnight build for that purpose.

The cache is per "build" (maybe we should call these "matrix rows" or something), i.e., the cache for `clang-tidy` is separate from the cache for `clang-3.8` or `asan`. A nightly build would need to be per matrix row, that requires two more configuration files, one in GitHub (probably an empty `$row-nightly.cfg` file) and one in the Kokoro config, analogous to `integration-nightly.cfg`.

That is just FYI, so if anybody starts working on it they don't waste as much time.

**Comment by mr-salty** [comment]
thanks... yeah, I was envisioning running every build nightly no-cache, although maybe the cost (quota / resources) vs benefit isn't worthwhile... I'm not entirely sure how to evaluate that.

**Comment by vnvo2409** [comment]
> I was envisioning running every build nightly no-cache

I think we could 
- open a PR with `--cache_test_results=no`,
- add `kokoro:run` to invalidate the cache.
- close the PR without further action.

The point is checking if there is any change in the behavior of production ( which is very unlikely to happen ). Combined with #5462, doing this process 2 or 3 times a year is enough for me. WDYT ?

**Comment by coryan** [comment]
> The point is checking if there is any change in the behavior of production ( which is very unlikely to happen ). Combined with #5462, doing this process 2 or 3 times a year is enough for me. WDYT ?

Your comment got me thinking, thanks: we do something very close to that once a month.  After every release we increase all the version numbers, that invalidated all the caches because the common library is modified when that happens.  Not sure that changes anything, just something I had not thought of.

**Comment by vnvo2409** [comment]
> After every release we increase all the version numbers, that invalidated all the caches because the common library is modified when that happens.

I've just had a look at #5399 and all the caches were invalidated indeed. Therefore, I think that this problem isn't really serious. We could eventually know if there is any change in production within a month. There is no need for invalidating the cache manually or periodically.

**Comment by coryan** [comment]
I think this was fixed in #5507, in particular:

https://github.com/googleapis/google-cloud-cpp/blob/e5171ad3e8322e11ed6b2798ae273ede59d4935f/google/cloud/storage/ci/run_integration_tests_emulator_bazel.sh#L87

and

https://github.com/googleapis/google-cloud-cpp/blob/e5171ad3e8322e11ed6b2798ae273ede59d4935f/google/cloud/storage/ci/run_integration_tests_emulator_bazel.sh#L92

There is a TODO left, but it is really only tangentially related, and getting fixed in #6886
